FR #1054: Initial commit with Requirements Model Data for FOCUS 1.2 (#1209)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: guidelines/writing-conformance-rules.md

Reviewer: Recommend naming this `writing-conformance-guidelines.md` to be in-line with the other markdowns in this folder, which all end with the term `guidelines`.

Author: renamed but kept this as a symbolic link to the new filename as this is linked in the open AI tickets. We can remove this link once the CR rule writing work is completed

Reviewer: Happy to resolve this now if you like!

Author: Ill leave it open as a reminder to clean up the symlink before we merge to working draft

> # Guidelines for writing confirmance rules

Reviewer: ```suggestion
# Guidelines for writing conformance rules
```

The `specification/conformance` folder contains modular CR components and a Python-based build process that assembles them into a validated `cr-<version>.json` file using a corresponding JSON Schema (`cr_schema.json`).

## Conformance document overview

The conformance document for FOCUS contains the following major sections:

| Section | Purpose |
|---------|---------|
| Details | Key details about the conformance document |
| ApplicabilityCriteria | Key flags used to define attributes about the data generator that need to be true for some conformance rules to apply |
| CheckFunctions | Method definitions to describe the actual check needed to conform to a rule |
| ConformanceDatasets | List of datasets defined by FOCUS and the related top level conformance rules associated with the dataset |
| ConformanceRules | Individual conformance rule definitions that are linked together by requirements and dependancies to form the full conformance ruleset |

## Steps to apply conformance rules to existing attributes and columns

> A Action Item (AI) ticket should be opened to track the progress of implementing the conformance rules for an existing check

Reviewer: ```suggestion
An Action Item (AI) ticket should be opened to track the progress of implementing the conformance rules for an existing check.
```

- `Description` - Free form description mostly used to help with tracking the work in stage 1
- `Reference` - The Column/Attribute Id this rule applies to
- `ApplicabilityCriteria` - Specific criteria that must be true of the data generator for this rule to apply to the dataset
- `Condition` - The definition of conditions under which this rule applies
- `Requirement` - The definition of what is required for conformance
- `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `RECOMMENDED`, `SHOULD`, `MAY`, `OPTIONAL`)
- `MustSatisfy` - The normative text that this rule defines
- `Type` - Identifier if this is a Static or Dynamic rule, with Static rules being possible to assess conformance without external information being required
- `Notes` - Free form notes (short) included in the conformance rule document
- `CRVersionIntroduced` - CR Version this rule was added to the Conformance Rules (See: [CR Versioning](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/1121-ai-align-on-approach-for-scrs/specification/conformance/README.md#versioning))
- `Status` - Status of the rule (Valid values: Active, Depreciated)

> ### Stage 2

Reviewer: Should we make some mention of Stage 3, even though it's out of scope for this exercise?

Also, is there a document we can/should link that describes all three stages?  I think @jpradocueva had a slide deck for this?  Or maybe we need to add a section that describes these stages all in one place?

Reviewer: Discussed in Sep 23 TF-CR.  @mike-finopsorg will review and confirm.

Author: This is covered in the README.md of the conformance folder.

> The second phase of conversion is to take the table created in Stage 1 and create the entires into the `specification/conformance` folder that adds the rules to the formal json structure.

Reviewer: ```suggestion
The second phase of conversion is to take the table created in Stage 1 and create the entries in the `specification/conformance` folder that adds the rules to the formal JSON structure.
```

- `conformance_rules/colunns/`: JSON files defining multiple `ConformanceRules` for a single column

#### Steps

- Assign the Action Item (AI) to yourself to signal that you are working on the item (See: [GitHub Issues](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues))
- Open a branch with the source of git branch `1121-ai-align-on-approach-for-scrs` for your development work with the naming format as follows `ai number`-cr-`entity-name`-1121. (example: 1255-cr-AvailabilityZone-1121)
- Pull your branch to your development environment and perform all work specific to this AI in this branch.
- Add a file into the relevant folder `conformance_rules/attributes/` or `conformance_rules/colunns/` with name `entity-name`.json (example: availabilityzone.json)
- Write your rules into this file based on the rules in the Stage 1 table from the AI ticket (See: [ConformanceRule Templates](#conformancerule-templates) for helpers)
- If you need to add new ApplicabilityCriteria add them to `applicability_criteria.json` avoiding duplication
- If you need to add new CheckFunctions add them to `check_functions.json` avoiding duplication

> - Add your top level conformance rule entry into the relevant Dataset entries in the `conformance_datasets.json` file

Reviewer: @mike-finopsorg I think this section needs a few more steps as discussed in the Aug 5 TF-CR meeting around how to structure the PR, how to flag you, and then what steps you want for review after that (e.g., Mike will review in the weekly meeting, or some other way that gets the PR approved and merged).

Reviewer: In particular, I want to make sure we capture the PR approval process.  I would like to not tie it to the Thu Members meeting, particularly because you cannot attend that meeting due to timezones, @mike-finopsorg.

I think you have license to approve and merge yourself.  Regardless, I'll leave it to you to describe what we need to document here for guidance.

Author: Added section. I will self approve into the working branch. It will all get reviewed by all members when this PR is brought out of draft to merge into `working_draft`

Reviewer 2: We can merge into the feature branch `1121-ai-align-on-approach-for-crs` and report on the Conformance Meeting on Tuesdays at 1 pm PT. Formal approval of the `working_draft` must be approved at the Member's meeting on Thursdays at 11:00 PT.

      "MustSatisfy": "",
      "Keyword": "MUST",
      "Requirement": {
        "CheckFunction": "AND",
        "Items": [
          {
            "CheckFunction": "CheckConformanceRule",
            "ConformanceRuleId": "SampleColumn-C-001-M"
          },
          {
            "CheckFunction": "CheckConformanceRule",
            "ConformanceRuleId": "SampleColumn-C-002-M"
          },
          {
            "CheckFunction": "CheckConformanceRule",
            "ConformanceRuleId": "SampleColumn-C-003-M"
          }
        ]
      },
      "Condition": {},
      "Dependencies": []
    }
  }
```

### Presence requirement rule

```json
  "SampleColumn-C-001-M": {
    "Function": "Presence",
    "Reference": "SampleColumn",
    "EntityType": "Column",
    "Notes": "",
    "CRVersionIntroduced": "1.2",
    "Status": "Active",
    "ApplicabilityCriteria": [],
    "Type": "Static",
    "ValidationCriteria": {
      "MustSatisfy": "MUST be present in a FOCUS dataset",
      "Keyword": "MUST",
      "Requirement": {
        "CheckFunction": "ColumnPresent",
        "ColumnName": "SampleColumn"
      },
      "Condition": {},
      "Dependencies": []
    }
  }
```

> ### Not NULL requirement rule
> 
> Common rule for columns with a not NULL requirement. Can also be used when there is a NOT NULL on condition.

Reviewer: ```suggestion
### NOT NULL requirement rule

Common rule for columns with a NOT NULL requirement. Can also be used when there is a NOT NULL condition.
```
